File: app/src/main/java/com/helpcrunch/demo/screens/SendMessageActivity.java
package com.helpcrunch.demo.screens;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.core.content.ContextCompat;
import androidx.core.content.res.ResourcesCompat;

import com.helpcrunch.demo.R;
import com.helpcrunch.library.core.HelpCrunch;

import java.util.HashMap;

public class SendMessageActivity extends AppCompatActivity {
    private final BroadcastReceiver hcEventsBroadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            HelpCrunch.Event event = (HelpCrunch.Event) intent.getSerializableExtra(HelpCrunch.EVENT_TYPE);
            HashMap<String, String> data = (HashMap<String, String>) intent.getSerializableExtra(HelpCrunch.EVENT_DATA);

            setSendButtonParameters(View.VISIBLE, View.GONE, true);

            if (event == null) {
                Log.w(HelpCrunch.EVENTS, "Can't receive data");
                return;
            }

            if (event == HelpCrunch.Event.MESSAGE_SENDING) {
                if (data != null) {
                    String error = data.get("error");
                    String resultData = data.get("data");

                    if (error != null) {
                        Toast.makeText(SendMessageActivity.this, error, Toast.LENGTH_SHORT).show();
                    } else {
                        addMessage(resultData);
                    }
                }
            }
        }
    };

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_send_message);

        initViews();

        registerReceiver(hcEventsBroadcastReceiver, new IntentFilter(HelpCrunch.EVENTS));
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        unregisterReceiver(hcEventsBroadcastReceiver);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            this.finish();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    private void initViews() {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(R.string.send_message);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        findViewById(R.id.send_button).setOnClickListener(v -> sendMessage());
    }

    private void sendMessage() {
        String text = ((EditText) findViewById(R.id.message_text)).getText().toString();

        if (TextUtils.isEmpty(text)) {
            Toast.makeText(SendMessageActivity.this, "Text is empty", Toast.LENGTH_SHORT).show();

            return;
        }

        boolean isForceNewChat = ((CheckBox) findViewById(R.id.force_new_chat)).isChecked();

        HelpCrunch.sendMessage(text, isForceNewChat);

        setSendButtonParameters(View.GONE, View.VISIBLE, false);
    }

    private void addMessage(String resultData) {
        LinearLayout messagesView = findViewById(R.id.messages);
        Drawable drawable = ResourcesCompat.getDrawable(getResources(), R.drawable.ic_hc_send, null);
        drawable.setColorFilter(ContextCompat.getColor(this, R.color.colorBlue), PorterDuff.Mode.SRC_IN);
        drawable.setBounds(0, 0, drawable.getIntrinsicWidth(), drawable.getIntrinsicHeight());

        AppCompatTextView textView = new AppCompatTextView(this);
        textView.setGravity(Gravity.CENTER_VERTICAL);
        textView.setCompoundDrawables(drawable, null, null, null);
        textView.setText(resultData);
        textView.setCompoundDrawablePadding(getResources().getDimensionPixelSize(R.dimen.messages_padding));
        messagesView.addView(textView);
    }

    private void setSendButtonParameters(int iconVisible, int progressVisible, boolean enabled) {
        findViewById(R.id.send_button_icon).setVisibility(iconVisible);
        findViewById(R.id.send_button_progress).setVisibility(progressVisible);
        findViewById(R.id.send_button).setEnabled(enabled);
    }
}

File: app/src/main/java/com/helpcrunch/demo/screens/SendMessageActivity.kt
package com.helpcrunch.demo.screens

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.PorterDuff
import android.graphics.PorterDuffColorFilter
import android.os.Bundle
import android.text.TextUtils
import android.text.style.ForegroundColorSpan
import android.util.Log
import android.view.Gravity
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatTextView
import androidx.core.content.ContextCompat
import androidx.core.content.res.ResourcesCompat
import androidx.core.text.buildSpannedString
import androidx.core.text.inSpans
import androidx.core.view.ViewCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.helpcrunch.demo.R
import com.helpcrunch.demo.databinding.ActivitySendMessageBinding
import com.helpcrunch.library.core.Callback
import com.helpcrunch.library.core.HelpCrunch

class SendMessageActivity : AppCompatActivity() {
    private lateinit var binding: ActivitySendMessageBinding

    private val hcEventsBroadcastReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val event = intent.getSerializableExtra(HelpCrunch.EVENT_TYPE) as HelpCrunch.Event?
            val data =
                intent.getSerializableExtra(HelpCrunch.EVENT_DATA) as HashMap<String, String>?
            binding.sendTextIcon.setLoading(false)
            binding.sendText.isEnabled = true

            if (event == null) {
                Log.w(HelpCrunch.EVENTS, "Can't receive data")
                return
            }

            if (event == HelpCrunch.Event.MESSAGE_SENDING) {
                if (data != null) {
                    val error = data["error"]
                    val resultData = data["data"]

                    if (error != null) {
                        addMessage(buildSpannedString {
                            inSpans(
                                ForegroundColorSpan(
                                    ResourcesCompat.getColor(
                                        resources,
                                        R.color.hc_color_red_error,
                                        theme
                                    )
                                )
                            ) {
                                append(error)
                            }
                        })
                    } else {
                        addMessage(resultData)
                    }
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySendMessageBinding.inflate(layoutInflater)
        setContentView(binding.root)

        initViews()
        LocalBroadcastManager.getInstance(this)
            .registerReceiver(hcEventsBroadcastReceiver, IntentFilter(HelpCrunch.EVENTS))
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onDestroy() {
        super.onDestroy()
        LocalBroadcastManager.getInstance(this).unregisterReceiver(hcEventsBroadcastReceiver)
    }

    private fun initViews() {
        ViewCompat.setOnApplyWindowInsetsListener(binding.root) { v, windowInsets ->
            val insets = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars())

            v.setPadding(
                insets.left,
                insets.top,
                insets.right,
                insets.bottom
            )

            WindowInsetsCompat.CONSUMED
        }

        WindowCompat.getInsetsController(window, window.decorView)
            .isAppearanceLightStatusBars = true

        setSupportActionBar(binding.toolbar)

        if (supportActionBar != null) {
            supportActionBar!!.setTitle(R.string.send_message)
            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        }

        binding.sendText.setOnClickListener { v: View? -> sendMessage() }
    }

    private fun sendMessage() {
        binding.sendTextIcon.setLoading(true)
        binding.sendText.isEnabled = false

        val text = binding.messageText.text.toString()

        if (TextUtils.isEmpty(text)) {
            Toast.makeText(this@SendMessageActivity, "Text is empty", Toast.LENGTH_SHORT).show()
            return
        }
        val isForceNewChat = binding.forceNewChat.isChecked
        HelpCrunch.sendMessage(text, isForceNewChat, object : Callback<String?>() {
            override fun onSuccess(result: String?) {
                Log.d("sendMessage + ", result.orEmpty())
            }

            override fun onError(message: String) {
                Log.d("sendMessage - ", message)
            }
        })
    }

    private fun addMessage(resultData: CharSequence?) {
        val drawable = ResourcesCompat.getDrawable(resources, R.drawable.ic_hc_send, null)
        drawable!!.colorFilter = PorterDuffColorFilter(
            ContextCompat.getColor(this, R.color.colorBlue),
            PorterDuff.Mode.SRC_IN
        )
        drawable.setBounds(0, 0, drawable.intrinsicWidth, drawable.intrinsicHeight)
        val textView = AppCompatTextView(this)
        textView.gravity = Gravity.CENTER_VERTICAL
        textView.setCompoundDrawables(drawable, null, null, null)
        textView.text = resultData
        textView.compoundDrawablePadding = resources.getDimensionPixelSize(R.dimen.messages_padding)
        binding.messages.addView(textView)
    }
}
